mantidproject/mantid
# Notify refl gui when region selector changed

This PR links up notifications from the region selector widget to the parent presenter on the ISIS Reflectometry GUI:
- Add a RegionSelectorObserver interface and expose this to python
- Add a subscribe method to the RegionSelector that accepts a RegionSelectorObserver and notifies it when the region is changed
- Inherit RegionSelectorObserver in the PreviewPresenter. When notified, it calls back to get the current region from the RegionSelector. For now it just prints the result to the log.

This is very basic functionality for now and is incomplete because it is part of a larger piece of work.

Note that I had to add a stub class to the PreviewPresenter to pass through as the observer because I could not easily get a shared_ptr to the PreviewPresenter itself, which caused problems when passing it through to a python object, which I don't think can be done with a raw pointer or reference. I'm not sure if there is a better way around this.

Part of #34105 

**To test:**

- Ensure this workspace is loaded or is in your load path: [INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/7504266/INTER45455_inst.zip)
- Open the ISIS Reflectometry GUI
- Select the Preview tab.
- Type in `INTER45455_inst`, set the angle to `0.5`, and click Load. The instrument view plot should display the data. Click the rectangle-select button above it and select a region. The segments will be summed and the result plotted on the second plot (spectrum vs TOF).
- Click the rectangle-selection button on the second plot and select a region. A message should be displayed in the log with the y range that was selected.

![image](https://user-images.githubusercontent.com/12895056/177389286-1c03a46b-2ac5-4261-a8c4-08d557536415.png)

*This does not require release notes* because **it is part of ongoing work that is not exposed to users**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
